pkg/kubelet/kubelet.go

```diff
@@ -1152,120 +1152,183 @@ func (kl *Kubelet) pullImageAndRunContainer(pod *api.BoundPod, container *api.Co
+				// killed them when restarting Infra Container.
+				if pod.Spec.RestartPolicy.Never == nil {
+					glog.V(1).Infof("Infra Container is being recreated. %q will be restarted.", container.Name)
+					containersToStart[index] = empty{}
 				}
-				containersInPod.RemoveContainerWithID(containerID)
-			}
-			containersInPod.RemoveContainerWithID(containerID)
-			if err := kl.killContainer(dockerContainer); err != nil {
-				glog.V(1).Infof("Failed to kill container %q: %v", dockerContainer.ID, err)
 				continue
 			}
+		} else {
+			if kl.shouldContainerBeRestarted(pod, container.Name, dockerContainerName, uid) {
+				// If we are here it means that the container is dead and sould be restarted, or never existed and should
+				// be created. We may be inserting this ID again if the container has changed and it has
+				// RestartPolicy::Always, but it's not a big deal.
+				glog.V(3).Infof("Container %+v is dead, but RestartPolicy says that we should restart it.", container)
+				containersToStart[index] = empty{}
+			}
 		}
+	}
 
-		if !kl.shouldContainerBeRestarted(pod, container.Name, dockerContainerName, uid) {
-			continue
-		}
+	// After the loop one of the following should be true:
+	// - createPodInfraContainer is true and containersToKeep is empty
+	// - createPodInfraContainer is false and containersToKeep contains at least ID of Infra Container
+
+	// If Infra container is the last running one, we don't want to keep it.
+	if !createPodInfraContainer && len(containersToStart) == 0 && len(containersToKeep) == 1 {
+		containersToKeep = make(map[dockertools.DockerID]int)
+	}
 
-		glog.V(3).Infof("Container with name %s doesn't exist, creating", dockerContainerName)
+	return podContainerChangesSpec{
+		StartInfraContainer: createPodInfraContainer,
+		InfraContainerId:    podInfraContainerID,
+		ContainersToStart:   containersToStart,
+		ContainersToKeep:    containersToKeep,
+	}, nil
+}
 
-		containerID, err := kl.pullImageAndRunContainer(pod, &container, &podVolumes, podInfraContainerID)
-		if err == nil {
-			containersInPod.RemoveContainerWithID(containerID)
+func (kl *Kubelet) syncPod(pod *api.BoundPod, containersInPod dockertools.DockerContainers) error {
+	podFullName := GetPodFullName(pod)
+	uid := pod.UID
+	containerChanges, err := kl.computePodContainerChanges(pod, containersInPod)
+	glog.V(3).Infof("Got container changes for pod %q: %+v", podFullName, containerChanges)
+	if err != nil {
+		return err
+	}
+
+	if containerChanges.StartInfraContainer || (len(containerChanges.ContainersToKeep) == 0 && len(containerChanges.ContainersToStart) == 0) {
+		glog.V(4).Infof("Killing Infra Container for %q", podFullName)
+		// Killing phase: if we want to start new infra container, or nothing is running kill everything (including infra container)
+		if podInfraContainer, found, _ := containersInPod.FindPodContainer(podFullName, uid, dockertools.PodInfraContainerName); found {
```

Comment:
Yeah, I thought about adding 'containersToKill' to changes struct to make everything explicit. This would move all this logic to the analysis part where it belong, but the cost is some information redundancy. I don't feel strongly about any one of those solutions...